emukit/quadrature/methods/warpings.py

```diff
@@ -0,0 +1,91 @@
+import numpy as np
+from typing import Optional
+
+
+class Warping:
+    """The warping for warped Bayesian quadrature."""
+
+    def transform(self, Y: np.ndarray) -> np.ndarray:
+        """Transform from base-GP to integrand.
+
+        :param Y: function values of latent function, shape (num_points, 1).
+        :return: transformed values, shape (num_points, 1)
+        """
+        raise NotImplemented
+
+    def inverse_transform(self, Y: np.ndarray) -> np.ndarray:
+        """Transform from integrand to base-GP.
+
+        :param Y: function values of integrand, shape (num_points, 1).
+        :return: transformed values, shape (num_points, 1)
+        """
+        raise NotImplemented
+
+    def update_parameters(self, **kargs) -> None:
+        """Update the warping parameters. Use `pass` if there are no parameters."""
+        raise NotImplementedError
+
+
+class IdentityWarping(Warping):
+
+    def transform(self, Y: np.ndarray) -> np.ndarray:
+        """Transform from base-GP to integrand.
+
+        :param Y: function values of latent function, shape (num_points, 1).
+        :return: transformed values, shape (num_points, 1)
+        """
+        return Y
+
+    def inverse_transform(self, Y: np.ndarray) -> np.ndarray:
+        """Transform from integrand to base-GP.
+
+        :param Y: function values of integrand, shape (num_points, 1).
+        :return: transformed values, shape (num_points, 1)
+        """
+        return Y
+
+    def update_parameters(self) -> None:
+        """No update for the identity transform."""
+        pass
+
+
+class SquareRootWarping(Warping):
+    """The square root warping"""
+
+    def __init__(self, offset: float, inverted: Optional[bool]=False):
+        """
+        :param offset: the offset of the warping
+        :param inverted: inverts the warping if ``True``. Default is ``False``.
+        """
+        self.offset = offset
+        self.inverted = inverted
+
+    def transform(self, Y: np.ndarray) -> np.ndarray:
+        """Transform from base-GP to integrand.
+
+        :param Y: function values of latent function, shape (num_points, 1).
+        :return: transformed values, shape (num_points, 1)
+        """
+        if self.inverted:
+            return self.offset - 0.5 * (Y * Y)
+        else:
+            return self.offset + 0.5 * (Y * Y)
+
+    def inverse_transform(self, Y: np.ndarray) -> np.ndarray:
+        """Transform from integrand to base-GP.
+
+        :param Y: function values of integrand, shape (num_points, 1).
+        :return: transformed values, shape (num_points, 1)
+        """
+        if self.inverted:
+            return np.sqrt(2. * (self.offset - Y))
+        else:
+            return np.sqrt(2. * (Y - self.offset))
+
+    def update_parameters(self, offset: Optional[float]=None) -> None:
```

Comment:
Will this actually be used anywhere? It's very specific (talk about mix and match).

Perhaps it would make sense to do a different thing here? Make the warping object immutable, if you need to update, do something like

```
new_warping = SquaredRootWarping(new_offset, original_warping.inverted)
```

Just a wild suggestion here

Reply:
Yes, it;s needed by WSABI-L and WSABI-M. Both of these methods require the warping to be data dependent, so the warping changes after a new datapoint is collected (example [here](https://github.com/EmuKit/emukit/pull/375/files#diff-34e3d9470acc26afa739a9cf396ac75ca15dc290975d58ac165e97ed87620e53R63) in the draft-PR). 

So the options seem to be
1. Create a new immutable warping instance after each new datapoint is collected. This happens possibly a couple of thousand times. 
2. Create 1 instance of the warping and update its parameter after each new datapoint (currently this I do). 

What is better you think?

Reply:
Some more context: The current implementation is specific in the sense that `update_parameters` in the warping classes gets the new value and sets the new value of its own warping parameters (in this case `offset`). The default in the parent class I set to `pass`.  This methods is always called when the data is set in the parent class of all warped BQ models  [here](https://github.com/EmuKit/emukit/pull/379/files#diff-2fb1abe80576c9608cae8ae66a69543320695bbd669ce450e169c9b49ad53bcdR114). 

In case a specific BQ model (which inherits from said warped BQ model) requires a data dependent warping, the computation of the new value is done in the BQ model and not the warping. Example for WSABI-L method [here](https://github.com/EmuKit/emukit/pull/375/files#diff-34e3d9470acc26afa739a9cf396ac75ca15dc290975d58ac165e97ed87620e53R63). 
So warping is not convoluted with the specific BQ model. I opted for the `update_parameters` method instead of a setter for `offset` exactly for that reason, i.e., whenever a method requires the warping to be data-dependent, we do not need to change the warping code, *nor* `set_data` in the BQ method base class.

About mix and match: you cannot mix and match in BQ in the sense that to each BQ model there belongs a specific warping class, but the warping class need know know which BQ model it is being used with. Does that make sense?

Reply:
So, both options you suggest would work, but I guess it would not make the warping code less specific. I am wondering in general though: Is it better to create new instances frequently, or rather create one and update/set its parameters frequently (in case it has any)?

Reply:
You're right, it's probably not worth the effort, these immutable instances don't help much.